File: lib/models/events/mouse/mouse_event.dart
import '../event.dart';

enum MouseBttn { left, right }

enum MouseEventType { move, press, release, scroll }

class MouseEvent extends Event {
  static const String eventType = 'mouse';
  final MouseEventType mouseEventType;
  num x;
  num y;

  MouseBttn? button;
  num? dx;
  num? dy;

  MouseEvent({
    required this.x,
    required this.y,
    required this.mouseEventType,
    this.dx,
    this.dy,
    this.button,
  }) : super(type: eventType);

  @override
  Map<String, dynamic> toJson() {
    return super.toJson()
      ..addAll({
        'x': x,
        'y': y,
        'dx': dx,
        'dy': dy,
        'mouseEventType': mouseEventType.name,
        'button': button?.name,
      });
  }

  @override
  MouseEvent.fromJson(super.json)
      : x = num.parse(json['x'].toString()),
        y = num.parse(json['y'].toString()),
        dx = num.tryParse(json['dx'].toString()),
        dy = num.tryParse(json['dy'].toString()),
        mouseEventType = MouseEventType.values
            .firstWhere((e) => e.name == json['mouseEventType'].toString()),
        button = MouseBttn.values
            .where((e) => e.name == json['button'].toString())
            .firstOrNull,
        super.fromJson();
}

File: python/models/mouse_event.py
from enum import Enum
from .event import Event


class MouseBttn(Enum):
    left = 0
    right = 1


class MouseEventType(Enum):
    move = 0
    press = 1
    release = 2
    scroll = 3


class MouseEvent(Event):
    eventType = "mouse"

    def __init__(
        self,
        mouseEventType: MouseEventType,
        x: float,
        y: float,
        button: MouseBttn | None = None,
        dx: float | None = None,
        dy: float | None = None,
    ):
        super().__init__(MouseEvent.eventType)
        self.mouseEventType = mouseEventType
        self.x = x
        self.y = y
        self.button = button
        self.dx = dx
        self.dy = dy

    def toJson(self) -> dict:
        return {
            **super().toJson(),
            "mouseEventType": self.mouseEventType.name,
            "x": self.x,
            "y": self.y,
            "button": self.button.name if self.button != None else None,
            "dx": self.dx,
            "dy": self.dy,
        }

    def __load(self, json: dict):
        super().fromJson(json)
        self.mouseEventType = MouseEventType[str(json["mouseEventType"])]
        self.x = json["x"]
        self.y = json["y"]
        self.button = MouseBttn[str(json["button"])] if json["button"] != None else None
        self.dx = json["dx"]
        self.dy = json["dy"]

    @staticmethod
    def fromJson(json: dict) -> "MouseEvent":
        event = MouseEvent(MouseEventType.move, 0, 0)
        event.__load(json)
        return event
